File: Offer 11. 旋转数组的最小数字.java
class Solution {
    public int minArray(int[] numbers) {
        int n = numbers.length;
        int target = binSearch(numbers, 0, n - 1);
        return numbers[target];
    }

    public int binSearch(int[] numbers, int low, int high) {
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (numbers[mid] > numbers[high]) {
                low = mid + 1; // 最小数字肯定不在mid这一侧，因此+1
            } else if (numbers[mid] < numbers[high]) {
                high = mid;
            } else {
                high -= 1; // 由于mid和high相同，排除high一定不会出错
            }
        }
        return low;
    }
}

File: Offer 11. 旋转数组的最小数字.py
class Solution:
    def minArray(self, numbers: List[int]) -> int:
        left, right = 0, len(numbers) - 1
        while left <= right:
            mid = (left + right) // 2
            if numbers[mid] < numbers[right]:
                # 这种情况mid可能是最终结果，不能去掉
                right = mid
            elif numbers[mid] > numbers[right]:
                # 这种情况mid一定不是最终结果，因此可以去掉
                left = mid + 1
            else:
                # 缩小范围
                right -= 1
        return numbers[left]